Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,782 @@
+            dispatch_sync(dispatch_get_main_queue(), ^{
+                auto guide = MainView.safeAreaLayoutGuide;
+                xrView = [[MTKView alloc] initWithFrame:guide.layoutFrame device:MetalDevice];
+                [MainView addSubview:xrView];
+                xrView.colorPixelFormat = MTLPixelFormatBGRA8Unorm;
+                xrView.depthStencilPixelFormat = MTLPixelFormatDepth32Float_Stencil8;
+                MetalLayer = (CAMetalLayer *)xrView.layer;
+                MetalLayer.device = MetalDevice;
+                auto scale = UIScreen.mainScreen.scale;
+                viewportSize.x = guide.layoutFrame.size.width * scale;
+                viewportSize.y = guide.layoutFrame.size.height * scale;
+            });
+
+            // Create and configure the ARKit session.
+            configuration = [ARWorldTrackingConfiguration new];
+            session = [ARSession new];
+            MetalDevice = id<MTLDevice>(graphicsContext);
+            sessionDelegate = [[SessionDelegate new]init:&ActiveFrameViews metalContext:MetalDevice];
+            session.delegate = sessionDelegate;
+            configuration.planeDetection = ARPlaneDetectionHorizontal;
+            configuration.lightEstimationEnabled = false;
+            configuration.worldAlignment = ARWorldAlignmentGravity;
+            [session runWithConfiguration:configuration];
+
+            // build pipeline
+            NSError* error;
+            id<MTLLibrary> lib = CompileShader(MetalDevice, shaderSource);
+            id<MTLFunction> vertexFunction = [lib newFunctionWithName:@"vertexShader"];
+            id<MTLFunction> fragmentFunction = [lib newFunctionWithName:@"fragmentShader"];
+
+            // Configure a pipeline descriptor that is used to create a pipeline state.
+            MTLRenderPipelineDescriptor *pipelineStateDescriptor = [[MTLRenderPipelineDescriptor alloc] init];
+            pipelineStateDescriptor.label = @"XR Pipeline";
+            pipelineStateDescriptor.vertexFunction = vertexFunction;
+            pipelineStateDescriptor.fragmentFunction = fragmentFunction;
+            pipelineStateDescriptor.colorAttachments[0].pixelFormat = MetalLayer.pixelFormat;
+
+            pipelineState = [MetalDevice newRenderPipelineStateWithDescriptor:pipelineStateDescriptor error:&error];
+            if (!pipelineState)
+            {
+                NSLog(@"Failed to create pipeline state: %@", error);
+            }
+            commandQueue = [MetalDevice newCommandQueue];
+            
+            // default fov values to avoid NaN at startup
+            auto& frameView = ActiveFrameViews[0];
+            frameView.FieldOfView.AngleDown = -(frameView.FieldOfView.AngleUp = 0.5);
+            frameView.FieldOfView.AngleLeft = -(frameView.FieldOfView.AngleRight = 0.5);
+        }
+
+        ~Impl()
+        {
+            [sessionDelegate release];
+            [configuration release];
+            [session release];
```

Comment:
Do you need to explicitly release `_cameraTextureY` and `_cameraTextureCbCr`?

Reply:
These should automatically be managed by the texture cache, I do need to free the MTLTexture which I added in my last iteration.